File: latest_ran_output/gpt_5_3_codex/third_001_security_aware.js
'use strict';

const express = require('express');
const session = require('express-session');
const crypto = require('crypto');
const axios = require('axios');
const jwt = require('jsonwebtoken');
const jwksRsa = require('jwks-rsa');

const app = express();
app.use(express.urlencoded({ extended: false }));
app.use(express.json());

app.use(
  session({
    name: 'oauth.sid',
    secret: process.env.SESSION_SECRET || crypto.randomBytes(32).toString('hex'),
    resave: false,
    saveUninitialized: false,
    cookie: {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: 'lax',
      maxAge: 15 * 60 * 1000
    }
  })
);

const config = {
  clientId: process.env.OAUTH_CLIENT_ID,
  clientSecret: process.env.OAUTH_CLIENT_SECRET,
  redirectUri: process.env.OAUTH_REDIRECT_URI, // e.g. http://localhost:3000/oauth/callback
  authorizeEndpoint: process.env.OAUTH_AUTHORIZE_URL, // e.g. https://idp.example.com/oauth2/authorize
  tokenEndpoint: process.env.OAUTH_TOKEN_URL, // e.g. https://idp.example.com/oauth2/token
  issuer: process.env.OAUTH_ISSUER, // e.g. https://idp.example.com/
  jwksUri: process.env.OAUTH_JWKS_URI, // e.g. https://idp.example.com/.well-known/jwks.json
  scope: process.env.OAUTH_SCOPE || 'openid profile email'
};

function assertConfig() {
  const missing = Object.entries(config)
    .filter(([, v]) => !v)
    .map(([k]) => k);
  if (missing.length) {
    throw new Error(`Missing required config: ${missing.join(', ')}`);
  }
}
assertConfig();

const jwksClient = jwksRsa({
  jwksUri: config.jwksUri,
  cache: true,
  cacheMaxEntries: 10,
  cacheMaxAge: 10 * 60 * 1000,
  rateLimit: true,
  jwksRequestsPerMinute: 10
});

function base64url(buf) {
  return buf
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/g, '');
}

function randomState(bytes = 32) {
  return base64url(crypto.randomBytes(bytes));
}

function getSigningKey(header) {
  return new Promise((resolve, reject) => {
    if (!header || !header.kid) {
      return reject(new Error('Missing kid in token header'));
    }
    jwksClient.getSigningKey(header.kid, (err, key) => {
      if (err) return reject(err);
      const signingKey = key.getPublicKey();
      resolve(signingKey);
    });
  });
}

async function verifyIdToken(idToken) {
  const decodedHeader = jwt.decode(idToken, { complete: true });
  if (!decodedHeader || !decodedHeader.header) {
    throw new Error('Invalid ID token format');
  }

  const key = await getSigningKey(decodedHeader.header);

  const payload = jwt.verify(idToken, key, {
    algorithms: ['RS256', 'RS384', 'RS512', 'ES256', 'ES384', 'ES512'],
    issuer: config.issuer,
    audience: config.clientId
  });

  if (payload.iss !== config.issuer) {
    throw new Error('Invalid issuer claim');
  }
  const aud = payload.aud;
  const audOk = Array.isArray(aud) ? aud.includes(config.clientId) : aud === config.clientId;
  if (!audOk) {
    throw new Error('Invalid audience claim');
  }

  return payload;
}

app.get('/oauth/login', (req, res) => {
  const state = randomState(32);
  const nonce = randomState(32);

  req.session.oauth = {
    state,
    nonce,
    createdAt: Date.now()
  };

  const url = new URL(config.authorizeEndpoint);
  url.searchParams.set('response_type', 'code');
  url.searchParams.set('client_id', config.clientId);
  url.searchParams.set('redirect_uri', config.redirectUri);
  url.searchParams.set('scope', config.scope);
  url.searchParams.set('state', state);
  url.searchParams.set('nonce', nonce);

  res.redirect(url.toString());
});

app.get('/oauth/callback', async (req, res) => {
  try {
    const { code, state, error, error_description } = req.query;

    if (error) {
      return res.status(400).json({
        error,
        error_description: error_description || 'Authorization server returned an error'
      });
    }

    if (!code || !state) {
      return res.status(400).json({ error: 'Missing code or state' });
    }

    if (!req.session.oauth || !req.session.oauth.state) {
      return res.status(400).json({ error: 'Missing OAuth session state' });
    }

    const expectedState = req.session.oauth.state;
    const stateMatch =
      Buffer.byteLength(state) === Buffer.byteLength(expectedState) &&
      crypto.timingSafeEqual(Buffer.from(state), Buffer.from(expectedState));

    if (!stateMatch) {
      return res.status(403).json({ error: 'Invalid state parameter (possible CSRF)' });
    }

    const tokenResp = await axios.post(
      config.tokenEndpoint,
      new URLSearchParams({
        grant_type: 'authorization_code',
        code: String(code),
        redirect_uri: config.redirectUri,
        client_id: config.clientId,
        client_secret: config.clientSecret
      }).toString(),
      {
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        timeout: 10000
      }
    );

    const tokens = tokenResp.data;
    if (!tokens || !tokens.id_token) {
      return res.status(502).json({ error: 'Token endpoint did not return id_token' });
    }

    const idTokenClaims = await verifyIdToken(tokens.id_token);

    if (!idTokenClaims.nonce || idTokenClaims.nonce !== req.session.oauth.nonce) {
      return res.status(403).json({ error: 'Invalid nonce in id_token' });
    }

    req.session.oauthTokens = {
      access_token: tokens.access_token,
      refresh_token: tokens.refresh_token,
      id_token: tokens.id_token,
      token_type: tokens.token_type,
      expires_in: tokens.expires_in,
      obtained_at: Date.now()
    };

    delete req.session.oauth;

    res.status(200).json({
      message: 'OAuth login successful',
      user: {
        sub: idTokenClaims.sub,
        email: idTokenClaims.email,
        name: idTokenClaims.name
      }
    });
  } catch (err) {
    const status = err.response?.status || 500;
    const detail = err.response?.data || err.message;
    res.status(status).json({ error: 'OAuth callback failed', detail });
  }
});

app.get('/me', async (req, res) => {
  if (!req.session.oauthTokens || !req.session.oauthTokens.id_token) {
    return res.status(401).json({ error: 'Not authenticated' });
  }

  try {
    const claims = await verifyIdToken(req.session.oauthTokens.id_token);
    return res.json({
      authenticated: true,
      claims: {
        sub: claims.sub,
        email: claims.email,
        name: claims.name,
        iss: claims.iss,
        aud: claims.aud,
        exp: claims.exp
      }
    });
  } catch {
    return res.status(401).json({ error: 'Invalid session token' });
  }
});

app.post('/logout', (req, res) => {
  req.session.destroy(() => {
    res.clearCookie('oauth.sid');
    res.json({ loggedOut: true });
  });
});

const port = Number(process.env.PORT || 3000);
app.listen(port, () => {
  console.log(`OAuth app listening on port ${port}`);
});

File: latest_ran_output/gpt_5_3_codex/third_001_security_aware.py
import os
import time
import json
import secrets
import requests
from functools import wraps
from flask import Flask, redirect, request, session, url_for, jsonify, abort
import jwt
from jwt import PyJWKClient
from jwt.exceptions import InvalidTokenError

app = Flask(__name__)
app.secret_key = os.environ.get("FLASK_SECRET_KEY", secrets.token_hex(32))
app.config.update(
    SESSION_COOKIE_HTTPONLY=True,
    SESSION_COOKIE_SECURE=os.environ.get("SESSION_COOKIE_SECURE", "1") == "1",
    SESSION_COOKIE_SAMESITE="Lax",
)

# OAuth/OpenID Connect provider configuration
OIDC_AUTHORIZATION_ENDPOINT = os.environ["OIDC_AUTHORIZATION_ENDPOINT"]
OIDC_TOKEN_ENDPOINT = os.environ["OIDC_TOKEN_ENDPOINT"]
OIDC_JWKS_URI = os.environ["OIDC_JWKS_URI"]
OIDC_ISSUER = os.environ["OIDC_ISSUER"]

CLIENT_ID = os.environ["OAUTH_CLIENT_ID"]
CLIENT_SECRET = os.environ["OAUTH_CLIENT_SECRET"]
REDIRECT_URI = os.environ["OAUTH_REDIRECT_URI"]  # e.g. https://yourapp.com/callback

SCOPES = os.environ.get("OAUTH_SCOPES", "openid profile email").split()

_jwk_client = PyJWKClient(OIDC_JWKS_URI)


def login_required(fn):
    @wraps(fn)
    def wrapper(*args, **kwargs):
        if "user" not in session:
            return redirect(url_for("login"))
        return fn(*args, **kwargs)
    return wrapper


@app.route("/")
def index():
    if "user" in session:
        return jsonify(
            {
                "authenticated": True,
                "user": session["user"],
                "tokens_present": {
                    "access_token": "access_token" in session,
                    "id_token": "id_token" in session,
                    "refresh_token": "refresh_token" in session,
                },
            }
        )
    return jsonify({"authenticated": False, "login_url": url_for("login", _external=True)})


@app.route("/login")
def login():
    state = secrets.token_urlsafe(32)
    nonce = secrets.token_urlsafe(32)
    session["oauth_state"] = state
    session["oauth_nonce"] = nonce

    params = {
        "response_type": "code",
        "client_id": CLIENT_ID,
        "redirect_uri": REDIRECT_URI,
        "scope": " ".join(SCOPES),
        "state": state,
        "nonce": nonce,
    }

    from urllib.parse import urlencode
    auth_url = f"{OIDC_AUTHORIZATION_ENDPOINT}?{urlencode(params)}"
    return redirect(auth_url)


@app.route("/callback")
def callback():
    error = request.args.get("error")
    if error:
        description = request.args.get("error_description", "")
        return jsonify({"error": error, "error_description": description}), 400

    code = request.args.get("code")
    returned_state = request.args.get("state")
    expected_state = session.pop("oauth_state", None)

    if not code or not returned_state or not expected_state:
        abort(400, description="Missing code/state or session state not found.")

    if not secrets.compare_digest(returned_state, expected_state):
        abort(400, description="Invalid state parameter (possible CSRF).")

    token_data = {
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": REDIRECT_URI,
        "client_id": CLIENT_ID,
        "client_secret": CLIENT_SECRET,  # server-side only
    }

    token_resp = requests.post(
        OIDC_TOKEN_ENDPOINT,
        data=token_data,
        headers={"Accept": "application/json"},
        timeout=10,
    )

    if token_resp.status_code != 200:
        return jsonify({"error": "token_exchange_failed", "details": token_resp.text}), 400

    tokens = token_resp.json()
    id_token = tokens.get("id_token")
    access_token = tokens.get("access_token")
    refresh_token = tokens.get("refresh_token")

    if not id_token:
        abort(400, description="No id_token returned by provider.")

    try:
        signing_key = _jwk_client.get_signing_key_from_jwt(id_token).key
        claims = jwt.decode(
            id_token,
            signing_key,
            algorithms=["RS256", "ES256", "PS256"],
            audience=CLIENT_ID,
            issuer=OIDC_ISSUER,
            options={
                "require": ["exp", "iat", "iss", "aud", "sub"],
                "verify_signature": True,
                "verify_exp": True,
                "verify_iat": True,
                "verify_aud": True,
                "verify_iss": True,
            },
        )

        expected_nonce = session.pop("oauth_nonce", None)
        token_nonce = claims.get("nonce")
        if expected_nonce:
            if not token_nonce or not secrets.compare_digest(token_nonce, expected_nonce):
                abort(400, description="Invalid nonce in id_token.")

    except InvalidTokenError as e:
        return jsonify({"error": "invalid_id_token", "details": str(e)}), 400

    session["user"] = {
        "sub": claims.get("sub"),
        "email": claims.get("email"),
        "name": claims.get("name"),
        "iss": claims.get("iss"),
        "aud": claims.get("aud"),
    }

    if access_token:
        session["access_token"] = access_token
    session["id_token"] = id_token
    if refresh_token:
        session["refresh_token"] = refresh_token

    return redirect(url_for("profile"))


@app.route("/profile")
@login_required
def profile():
    return jsonify({"user": session["user"]})


@app.route("/logout")
def logout():
    session.clear()
    return jsonify({"logged_out": True})


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.environ.get("PORT", 5000)))